File: backend/app/schemas/agent/agent_execution.py
"""AgentExecution 查询接口 Pydantic 响应模型。

仅做响应结构定义，不含任何业务逻辑（聚合/JOIN 下沉到 service）。
"""
from __future__ import annotations

import json
from typing import Optional, List, Dict, Any
from datetime import datetime

from pydantic import BaseModel


# ── 单个执行记录项（列表用，输出截断） ──────────────────────────────────────

class AgentExecutionItem(BaseModel):
    """Agent 执行记录列表项。"""
    execution_id: str
    session_id: Optional[int] = None
    session_title: Optional[str] = None
    user_id: Optional[int] = None
    execution_mode: str
    target_id: Optional[str] = None
    status: str
    user_input: Optional[str] = None
    output: Optional[str] = None
    error: Optional[str] = None
    latency_ms: Optional[int] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ── 列表响应（分页） ────────────────────────────────────────────────────────

class AgentExecutionListResp(BaseModel):
    """执行记录分页列表响应。"""
    items: List[AgentExecutionItem]
    total: int
    page: int
    page_size: int


# ── 详情（含完整输出 + trace 概要） ──────────────────────────────────────────

class AgentTraceSummary(BaseModel):
    """关联链路概要。"""
    trace_id: Optional[str] = None
    agent_id: Optional[str] = None
    agent_type: Optional[str] = None
    status: Optional[str] = None
    duration_ms: Optional[int] = None

    model_config = {"from_attributes": True}


# ── 调用链节点事件（详情/链路用） ──────────────────────────────────────────

class AgentExecutionEventItem(BaseModel):
    """执行链节点事件：用于绘制分层 DAG 拓扑与时间线。

    拓扑相关字段（node_key / agent_code / role_name / node_status / layer /
    parent_keys / channel / node_type 等）存放于 metadata_json，这里提取为
    便捷字段便于前端直接使用。
    """
    id: int
    execution_id: str
    event_type: Optional[str] = None
    sequence: Optional[int] = None
    content: Optional[Any] = None
    source: Optional[str] = None
    source_id: Optional[str] = None
    # 拓扑便捷字段（来自 metadata_json）
    node_key: Optional[str] = None
    agent_code: Optional[str] = None
    role_name: Optional[str] = None
    node_status: Optional[str] = None
    node_type: Optional[str] = None
    channel: Optional[str] = None
    layer: Optional[int] = None
    parent_keys: List[str] = []
    created_at: Optional[datetime] = None
    metadata_json: Optional[Dict[str, Any]] = None

    model_config = {"from_attributes": True}

    @classmethod
    def from_orm_event(cls, event: Any) -> "AgentExecutionEventItem":
        meta = event.event_metadata or {}
        content = event.content
        if isinstance(content, (dict, list)):
            content = json.dumps(content, ensure_ascii=False)
        return cls(
            id=event.id,
            execution_id=event.execution_id,
            event_type=event.event_type,
            sequence=event.sequence,
            content=content,
            source=event.source,
            source_id=event.source_id,
            node_key=meta.get("node_key") or meta.get("node_role"),
            agent_code=meta.get("agent_code"),
            role_name=meta.get("role_name"),
            node_status=meta.get("node_status"),
            node_type=meta.get("node_type"),
            channel=meta.get("channel"),
            layer=meta.get("layer"),
            parent_keys=meta.get("parent_keys") or [],
            created_at=event.created_at,
            metadata_json=meta,
        )


# ── 人工介入暂停（详情用） ─────────────────────────────────────────────────

class AgentHitlPauseItem(BaseModel):
    """人工介入暂停记录。"""
    id: int
    execution_id: str
    pause_type: Optional[str] = None
    status: Optional[str] = None
    payload: Optional[Dict[str, Any]] = None
    response_data: Optional[Dict[str, Any]] = None
    user_id: Optional[int] = None
    approver_id: Optional[int] = None
    comment: Optional[str] = None
    created_at: Optional[datetime] = None
    resolved_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ── 事件列表响应 ───────────────────────────────────────────────────────────

class AgentExecutionEventListResp(BaseModel):
    """调用链事件列表响应。"""
    items: List[AgentExecutionEventItem]
    total: int


class AgentExecutionDetail(AgentExecutionItem):
    """执行记录详情：在 Item 基础上含完整 output/error/metadata 与 trace 概要。"""
    metadata_json: Optional[Dict[str, Any]] = None
    trace: Optional[AgentTraceSummary] = None
    events: List[AgentExecutionEventItem] = []
    hitl_pauses: List[AgentHitlPauseItem] = []


# ── 统计聚合 ────────────────────────────────────────────────────────────────

class AgentExecutionStats(BaseModel):
    """执行记录统计聚合。"""
    total: int
    by_execution_mode: Dict[str, int]
    by_status: Dict[str, int]
    total_latency_ms: int
    avg_latency_ms: float

File: frontend/src/api/agentExecution.ts
import { getToken } from '@/utils/auth'

const baseUrl = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'

export type ExecutionMode = 'dify' | 'skill' | 'agent' | 'agent_team' | 'sqlbot'
export type ExecutionStatus = 'running' | 'completed' | 'failed'

export interface AgentExecutionItem {
  execution_id: string
  session_id: number
  session_title?: string
  user_id?: number
  execution_mode: ExecutionMode
  target_id?: string
  status: ExecutionStatus
  user_input?: string
  output?: string
  error?: string
  latency_ms?: number
  started_at?: string
  completed_at?: string
}

export interface AgentExecutionEvent {
  id: number
  execution_id: string
  event_type?: string
  sequence?: number
  content?: string
  source?: string
  source_id?: string
  // 拓扑便捷字段（来自 metadata_json）
  node_key?: string
  agent_code?: string
  role_name?: string
  node_status?: string
  node_type?: string
  channel?: string
  layer?: number
  parent_keys?: string[]
  created_at?: string
  metadata_json?: Record<string, any>
}

export interface AgentHitlPause {
  id: number
  execution_id: string
  pause_type?: string
  status?: string
  payload?: Record<string, any>
  response_data?: Record<string, any>
  user_id?: number
  approver_id?: number
  comment?: string
  created_at?: string
  resolved_at?: string
}

export interface AgentExecutionDetail extends AgentExecutionItem {
  metadata_json?: Record<string, any>
  trace?: Record<string, any>
  events?: AgentExecutionEvent[]
  hitl_pauses?: AgentHitlPause[]
}

export interface AgentExecutionEventListResp {
  items: AgentExecutionEvent[]
  total: number
}

export interface AgentExecutionListResp {
  items: AgentExecutionItem[]
  total: number
  page: number
  page_size: number
}

export interface AgentExecutionStats {
  by_execution_mode: Record<string, number>
  by_status: Record<string, number>
  avg_latency_ms?: number
  total_latency_ms?: number
  total?: number
  [key: string]: unknown
}

export interface ListAgentExecutionsParams {
  session_id?: number
  execution_mode?: ExecutionMode
  status?: ExecutionStatus
  target_id?: string
  keyword?: string
  start_time?: string
  end_time?: string
  page?: number
  page_size?: number
}

async function authHeader(): Promise<Record<string, string>> {
  const token = getToken()
  return token ? { Authorization: `Bearer ${token}` } : {}
}

function buildQuery(params: ListAgentExecutionsParams): string {
  const query = new URLSearchParams()
  if (params.page) query.set('page', String(params.page))
  if (params.page_size) query.set('page_size', String(params.page_size))
  if (params.session_id) query.set('session_id', String(params.session_id))
  if (params.execution_mode) query.set('execution_mode', params.execution_mode)
  if (params.status) query.set('status', params.status)
  if (params.target_id) query.set('target_id', params.target_id)
  if (params.keyword) query.set('keyword', params.keyword)
  if (params.start_time) query.set('start_time', params.start_time)
  if (params.end_time) query.set('end_time', params.end_time)
  return query.toString()
}

export async function listAgentExecutions(
  params: ListAgentExecutionsParams = {},
): Promise<AgentExecutionListResp> {
  const qs = buildQuery(params)
  const res = await fetch(`${baseUrl}/api/v1/agent-execution/?${qs}`, {
    headers: { ...(await authHeader()) },
  })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return res.json()
}

export async function getAgentExecutionDetail(
  executionId: string,
): Promise<AgentExecutionDetail> {
  const res = await fetch(`${baseUrl}/api/v1/agent-execution/${executionId}`, {
    headers: { ...(await authHeader()) },
  })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return res.json()
}

export async function listAgentExecutionsBySession(
  sessionId: number | string,
): Promise<AgentExecutionListResp> {
  const res = await fetch(`${baseUrl}/api/v1/agent-execution/session/${sessionId}`, {
    headers: { ...(await authHeader()) },
  })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return res.json()
}

export async function getAgentExecutionStats(): Promise<AgentExecutionStats> {
  const res = await fetch(`${baseUrl}/api/v1/agent-execution/stats`, {
    headers: { ...(await authHeader()) },
  })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return res.json()
}

export async function getAgentExecutionEvents(
  executionId: string,
): Promise<AgentExecutionEventListResp> {
  const res = await fetch(`${baseUrl}/api/v1/agent-execution/${executionId}/events`, {
    headers: { ...(await authHeader()) },
  })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return res.json()
}

/** 归一化事件回放：仅返回 step / artifact 两类统一事件（结构化 dict） */
export async function getUnifiedEvents(
  executionId: string,
): Promise<{ execution_id: string; steps: any[]; artifacts: any[]; total: number }> {
  const res = await fetch(`${baseUrl}/api/v1/agent-execution/${executionId}/unified-events`, {
    headers: { ...(await authHeader()) },
  })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return res.json()
}
